File: 백준/Silver/4949. 균형잡힌 세상/균형잡힌 세상.kt
import java.io.BufferedReader
import java.io.InputStreamReader
import java.util.*

fun main() {
    val br = BufferedReader(InputStreamReader(System.`in`))
    val sb = StringBuilder()

    while (true) {
        val sentence = br.readLine()
        val stack = Stack<Char>()

        if (sentence == ".") {
            break
        } else {
            for (char in sentence) {
                when (char) {
                    '(', '[' -> stack.push(char)
                    ')' -> {
                        if (stack.isNotEmpty() && stack.peek() == '(') {
                            stack.pop()
                        } else {
                            stack.push(char) // Mismatched closing bracket
                            break
                        }
                    }
                    ']' -> {
                        if (stack.isNotEmpty() && stack.peek() == '[') {
                            stack.pop()
                        } else {
                            stack.push(char) // Mismatched closing bracket
                            break
                        }
                    }
                }
            }
        }

        if (stack.isEmpty()) {
            sb.append("yes\n")
        } else {
            sb.append("no\n")
        }
    }

    println(sb)
}

File: 백준/Silver/4949. 균형잡힌 세상/균형잡힌 세상.py
from sys import stdin as s
while True:
    arr = list(map(str,s.readline().rstrip()))
    stack = []
    if arr[0] == '.':
        break
    count = 0
    for i in range(len(arr)):
       if arr[i] == '[':
           stack.append(arr[i])
       elif arr[i] == '(':
           stack.append(arr[i])
       elif arr[i] == ']':
           if stack:
            if stack.pop() == '[':
                continue
            else:
                count += 1
                break
           else:
               count += 1
       elif arr[i] == ')':
           if stack:
            if stack.pop() =='(':
                continue
            else: 
                count += 1
                break
           else:
               count += 1
       else:
           continue
    
    if count != 0:
        print('no')
    else:
        if stack:
            print('no')
        else:
            print('yes')
    